File: ui/src/api/status.ts
// Canonical spec lifecycle statuses in board order.
// Keep in sync with internal/spec/status.go (SpecStatuses).
export const SPEC_COLUMNS = [
  "draft",
  "planned",
  "in_progress",
  "in_review",
  "complete",
] as const;

export type SpecColumn = (typeof SPEC_COLUMNS)[number];

export const UNASSIGNED_COLUMN = "unassigned";

// Legacy/renamed statuses mapped to their current value.
// Keep in sync with internal/spec/status.go (legacyStatusAliases).
const LEGACY_STATUS_ALIASES: Record<string, string> = {
  refined: "planned",
  initial: "draft",
};

export function normalizeSpecStatus(status: string): string {
  const s = status.trim().toLowerCase();
  return LEGACY_STATUS_ALIASES[s] ?? s;
}

// columnForStatus normalizes legacy statuses first, then maps anything unknown
// to the Unassigned column.
export function columnForStatus(status: string): string {
  const s = normalizeSpecStatus(status);
  return (SPEC_COLUMNS as readonly string[]).includes(s) ? s : UNASSIGNED_COLUMN;
}

File: internal/spec/status.go
package spec

import "strings"

// UnassignedColumn is the board column for specs whose status is empty or not
// a known lifecycle status.
const UnassignedColumn = "unassigned"

// specStatuses is the canonical, ordered list of spec lifecycle statuses.
// Keep in sync with ui/src/api/status.ts (SPEC_COLUMNS).
var specStatuses = []string{
	"draft",
	"planned",
	"in_progress",
	"in_review",
	"complete",
}

// legacyStatusAliases maps removed/renamed statuses to their current value.
var legacyStatusAliases = map[string]string{
	"refined": "planned",
	"initial": "draft",
}

// SpecStatuses returns the canonical spec lifecycle statuses in board order.
func SpecStatuses() []string {
	out := make([]string, len(specStatuses))
	copy(out, specStatuses)
	return out
}

// NormalizeSpecStatus trims and lowercases a raw status and maps legacy values
// (refined → planned, initial → draft) to the current vocabulary. Unknown
// values pass through unchanged so callers can route them to Unassigned.
func NormalizeSpecStatus(status string) string {
	s := strings.ToLower(strings.TrimSpace(status))
	if mapped, ok := legacyStatusAliases[s]; ok {
		return mapped
	}
	return s
}

// ColumnForSpecStatus returns the board column for a raw status. It normalizes
// legacy values first; anything not in the canonical list maps to Unassigned.
func ColumnForSpecStatus(status string) string {
	s := NormalizeSpecStatus(status)
	for _, known := range specStatuses {
		if s == known {
			return s
		}
	}
	return UnassignedColumn
}
